File: src/facilitation/pacing.py
"""Pacing and turn-taking logic for meditation facilitation.

This is the most nuanced component - distinguishing between:
- Thinking pause (2-5 sec): Wait silently
- Contemplative dropping-in (5-30 sec): Wait silently
- Natural end of sharing (3-5 sec + falling intonation): Respond
- LLM-detected silence intent ([HOLD] signal): Enter extended listening mode
- Very long silence (120+ sec): Gentle check-in (optional)
"""

import time
from enum import Enum, auto

from ..config import PacingConfig


class ConversationState(Enum):
    """Current state of the conversation."""

    IDLE = auto()  # Session not started
    LISTENING = auto()  # Actively listening to meditator
    PROCESSING = auto()  # Processing what was said
    RESPONDING = auto()  # Facilitator is speaking
    SILENT_HOLD = auto()  # Extended silence mode (meditator requested)


class TurnDecision(Enum):
    """Decision about whether to take a turn."""

    WAIT = auto()  # Continue waiting
    RESPOND = auto()  # Time to respond
    CHECK_IN = auto()  # Gentle check-in after long silence
    HOLD = auto()  # In silence mode, keep holding


class PacingController:
    """Controls turn-taking dynamics in meditation facilitation.

    Handles the nuanced timing of when the facilitator should speak,
    respecting the meditator's contemplative process.
    """

    def __init__(self, config: PacingConfig | None = None):
        self.config = config or PacingConfig()

        self._state = ConversationState.IDLE
        self._last_speech_end: float = 0
        self._last_response_time: float = 0
        self._silence_mode_start: float | None = None
        self._check_in_count: int = 0
        self._has_spoken: bool = False

    @property
    def state(self) -> ConversationState:
        """Current conversation state."""
        return self._state

    def start_session(self) -> None:
        """Start a new meditation session."""
        self._state = ConversationState.LISTENING
        self._last_speech_end = 0
        self._last_response_time = time.time()
        self._silence_mode_start = None
        self._check_in_count = 0
        self._has_spoken = False

    def end_session(self) -> None:
        """End the current session."""
        self._state = ConversationState.IDLE

    def on_speech_start(self) -> None:
        """Called when meditator starts speaking."""
        self._state = ConversationState.LISTENING
        self._check_in_count = 0

    def on_speech_end(self) -> None:
        """Called when meditator stops speaking."""
        self._last_speech_end = time.time()
        self._state = ConversationState.PROCESSING
        self._has_spoken = True

    def on_transcription(self, text: str) -> TurnDecision:
        """Process transcribed text and decide on turn-taking.

        If in silence mode, any speech auto-exits it. The LLM decides
        whether to *enter* silence mode via the [HOLD] signal — that
        is handled externally after the LLM response is received.

        Args:
            text: Transcribed speech

        Returns:
            TurnDecision indicating what to do
        """
        if self._silence_mode_start is not None:
            self.exit_silence_mode()

        return TurnDecision.RESPOND

    def should_respond(self) -> TurnDecision:
        """Check if it's time to respond based on timing.

        Call this periodically during silence to check timing-based decisions.

        Returns:
            TurnDecision indicating what to do
        """
        now = time.time()

        # Exponential backoff: e.g. 240s, 480s, 960s, ...
        check_in_threshold = self.config.extended_silence_sec * (2 ** self._check_in_count)

        # If in silence mode, hold indefinitely (no check-ins)
        if self._silence_mode_start is not None:
            return TurnDecision.HOLD

        # Normal mode - check if enough time has passed since speech ended
        if self._last_speech_end > 0:
            silence_duration = now - self._last_speech_end
            response_delay = self.config.response_delay_ms / 1000.0

            if silence_duration >= response_delay:
                return TurnDecision.RESPOND

        # Check for extended silence in normal mode (only after conversation has started)
        if self._has_spoken:
            time_since_response = now - self._last_response_time
            if time_since_response >= check_in_threshold:
                return TurnDecision.CHECK_IN

        return TurnDecision.WAIT

    def on_check_in(self) -> None:
        """Record that a check-in occurred (increases backoff interval)."""
        self._check_in_count += 1

    def on_response_start(self) -> None:
        """Called when facilitator starts responding."""
        self._state = ConversationState.RESPONDING

    def on_response_end(self) -> None:
        """Called when facilitator finishes responding."""
        self._state = ConversationState.LISTENING
        self._last_response_time = time.time()
        self._last_speech_end = 0  # Reset for next turn

    def enter_silence_mode(self) -> None:
        """Enter extended silence mode (called after LLM returns [HOLD])."""
        self._state = ConversationState.SILENT_HOLD
        self._silence_mode_start = time.time()

    def exit_silence_mode(self) -> None:
        """Exit silence mode (called when meditator speaks again)."""
        self._state = ConversationState.LISTENING
        self._silence_mode_start = None

    def get_silence_duration(self) -> float:
        """Get current silence duration in seconds."""
        if self._silence_mode_start:
            return time.time() - self._silence_mode_start
        if self._last_speech_end > 0:
            return time.time() - self._last_speech_end
        return time.time() - self._last_response_time

    def is_in_silence_mode(self) -> bool:
        """Check if in extended silence mode."""
        return self._silence_mode_start is not None

File: ts/src/facilitation/pacing.ts
/**
 * Pacing and turn-taking logic for meditation facilitation.
 *
 * TS port of src/facilitation/pacing.py — see that file for the rationale
 * behind the distinctions between thinking pauses, contemplative drops,
 * natural end of sharing, and silence-mode requested via [HOLD].
 */

import { realClock, type Clock } from '../clock.js';

export const ConversationState = {
    Idle: 'idle',
    Listening: 'listening',
    Processing: 'processing',
    Responding: 'responding',
    SilentHold: 'silent_hold',
} as const;
export type ConversationState =
    (typeof ConversationState)[keyof typeof ConversationState];

export const TurnDecision = {
    Wait: 'wait',
    Respond: 'respond',
    CheckIn: 'check_in',
    Hold: 'hold',
} as const;
export type TurnDecision =
    (typeof TurnDecision)[keyof typeof TurnDecision];

export interface PacingConfig {
    // -----------------------------------------------------------------
    // Facilitation-side pacing — used by PacingController itself.
    // -----------------------------------------------------------------

    /** Milliseconds of silence after speech before responding. */
    responseDelayMs: number;
    /** Seconds of total silence before a gentle check-in. */
    silenceCheckinSec: number;
    /** If false, check-ins never fire. */
    silenceCheckinsEnabled: boolean;
    /**
     * If false, the [HOLD] signal from the LLM is ignored — the session
     * never enters extended silence mode. Useful for users who find
     * silence mode unsettling. The PacingController doesn't enforce
     * this on its own; callers should treat a hold-signaled response
     * as a normal one when this is false.
     */
    silenceModeEnabled: boolean;

    // -----------------------------------------------------------------
    // Client-side VAD tuning — used by STT adapters, not PacingController.
    // Grouped here (matching the Python `PacingConfig`) so the user has
    // one knob bag to tune. Adapters read only the fields they need.
    // -----------------------------------------------------------------

    /** Base trailing silence before submitting a transcribed utterance. */
    silenceBaseMs: number;
    /** Maximum tolerated silence after a long share. */
    silenceMaxMs: number;
    /** Extra ms of silence allowed per ms of speech (ramp from base to max). */
    silenceRampRate: number;
    /** Minimum total speech duration before an utterance can be submitted. */
    minSpeechDurationMs: number;
}

export const defaultPacingConfig: PacingConfig = {
    responseDelayMs: 2000,
    silenceCheckinSec: 300,
    silenceCheckinsEnabled: true,
    silenceModeEnabled: true,
    silenceBaseMs: 3000,
    silenceMaxMs: 5000,
    silenceRampRate: 0.12,
    minSpeechDurationMs: 500,
};

export interface PacingControllerOptions {
    config?: Partial<PacingConfig>;
    clock?: Clock;
}

export class PacingController {
    readonly config: PacingConfig;
    private readonly clock: Clock;

    private _state: ConversationState = ConversationState.Idle;
    private _lastSpeechEnd = 0;
    private _lastResponseTime = 0;
    private _silenceModeStart: number | null = null;
    private _hasSpoken = false;

    constructor(options: PacingControllerOptions = {}) {
        this.config = { ...defaultPacingConfig, ...options.config };
        this.clock = options.clock ?? realClock;
    }

    get state(): ConversationState {
        return this._state;
    }

    startSession(): void {
        this._state = ConversationState.Listening;
        this._lastSpeechEnd = 0;
        this._lastResponseTime = this.clock();
        this._silenceModeStart = null;
        this._hasSpoken = false;
    }

    endSession(): void {
        this._state = ConversationState.Idle;
    }

    onSpeechStart(): void {
        this._state = ConversationState.Listening;
    }

    onSpeechEnd(): void {
        this._lastSpeechEnd = this.clock();
        this._state = ConversationState.Processing;
        this._hasSpoken = true;
    }

    /**
     * Process a transcribed utterance and decide on turn-taking.
     *
     * Any speech auto-exits silence mode. Entering silence mode is the
     * LLM's call, handled externally via the [HOLD] signal in its reply.
     */
    onTranscription(_text: string): TurnDecision {
        if (this._silenceModeStart !== null) {
            this.exitSilenceMode();
        }
        return TurnDecision.Respond;
    }

    /**
     * Timing-based decision, called periodically during silence.
     */
    shouldRespond(): TurnDecision {
        const now = this.clock();

        if (this._silenceModeStart !== null) {
            return TurnDecision.Hold;
        }

        if (this._lastSpeechEnd > 0) {
            const silenceDuration = now - this._lastSpeechEnd;
            const responseDelay = this.config.responseDelayMs / 1000;
            if (silenceDuration >= responseDelay) {
                return TurnDecision.Respond;
            }
        }

        if (
            this._hasSpoken &&
            this.config.silenceCheckinsEnabled &&
            now - this._lastResponseTime >= this.config.silenceCheckinSec
        ) {
            return TurnDecision.CheckIn;
        }

        return TurnDecision.Wait;
    }

    onResponseStart(): void {
        this._state = ConversationState.Responding;
    }

    onResponseEnd(): void {
        this._state = ConversationState.Listening;
        this._lastResponseTime = this.clock();
        this._lastSpeechEnd = 0;
    }

    /** Called after the LLM signals [HOLD]. */
    enterSilenceMode(): void {
        this._state = ConversationState.SilentHold;
        this._silenceModeStart = this.clock();
    }

    /** Called when the meditator speaks again. */
    exitSilenceMode(): void {
        this._state = ConversationState.Listening;
        this._silenceModeStart = null;
    }

    getSilenceDuration(): number {
        const now = this.clock();
        if (this._silenceModeStart !== null) {
            return now - this._silenceModeStart;
        }
        if (this._lastSpeechEnd > 0) {
            return now - this._lastSpeechEnd;
        }
        return now - this._lastResponseTime;
    }

    isInSilenceMode(): boolean {
        return this._silenceModeStart !== null;
    }

    // Internal hooks used by tests that need to set up timing scenarios
    // without driving the controller through a full transcript. Mirrors
    // the Python tests reaching into _last_speech_end etc.
    /** @internal */
    _setLastSpeechEnd(t: number): void { this._lastSpeechEnd = t; }
    /** @internal */
    _setLastResponseTime(t: number): void { this._lastResponseTime = t; }
    /** @internal */
    _setHasSpoken(v: boolean): void { this._hasSpoken = v; }
}
